src/NHibernate/Collection/Generic/SetHelpers/SetSnapShot.cs

```diff
@@ -8,6 +8,22 @@ namespace NHibernate.Collection.Generic.SetHelpers
 	internal class SetSnapShot<T> : ISetSnapshot<T>
 	{
 		private readonly List<T> _elements;
```

Comment:
Changing this into a `LinkedHashMap<T, T>` (from `NHibernate.Util` namespace) could do most of the job, and that would be way simpler, also avoiding having two collections instead of one.
But it does not support a `null` key. It could be modified to have an option for supporting it, maybe. This option would cause it to no more respect `IDictionary<TKey, TValue>` contract which explicitly states `null` keys are not supported, so it is debatable.

Reply:
In fact we do not need the ordering preserving capability either for the `SetSnapshot`. Set are not order sensitive, included in dirty check logics.

So we could use a regular `Dictionary<T, T>` instead of the `List`, with your additional field for handling the `null` case. The enumerator would have to be overloaded for also yielding the null if there is some, the count too for taking the null element into account, ...

Reply:
My implementation focuses primarily on preventing any possible regressions as I am not so familiar with NH to anticipate what can and cannot appear in the snapshot. Other than being significantly more complex, LinkedHashMap would also not support multiple entries which would claim to be Equal. I assumed there was a reason for the List<T> based implementation of the SetSnapShot, it just failed terribly due to the way TryGetValue is used in a loop against a similarly sized collection.

Reply:
This snapshot helper class ends up being consumed only in `PersistentGenericSet`. It may by passed around in NHibernate code but only for being used inside some `PersistentGenericSet` instance.
So you can check what features it actually needs.
And as it is an internal class, it can be freely changed.

Reply:
There is no strong guarantee of input. Also, an instance of SetSnapShot gets passed to and returned from GetOrphans, used in SequenceEquals, etc. Scenarios where the lookup by key does not even come into play, that's why I am creating the lookup dictionary very lazily.
I would need extremely deep knowledge of NH to feel confident changing the (even undocumented) contract of a class that's been there unchanged since 2.0.